src/ophyd_async/panda/writers/hdf_writer.py

```diff
@@ -0,0 +1,234 @@
+import asyncio
+import atexit
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any, AsyncIterator, Dict, List, Optional, Union
+
+from bluesky.protocols import Asset, Descriptor, Hints
+from p4p.client.thread import Context
+
+from ophyd_async.core import (
+    DEFAULT_TIMEOUT,
+    DetectorWriter,
+    Device,
+    DirectoryProvider,
+    NameProvider,
+    SignalR,
+    SignalRW,
+    wait_for_value,
+)
+from ophyd_async.panda.panda import PandA
+
+from .panda_hdf import _HDFDataset, _HDFFile
+
+
+@dataclass()
+class HdfSignals:
+    file_path: SignalRW  # This is the directory rather than path. Path is read only
+    file_name: SignalRW
+    num_capture: SignalRW
+    num_captured: SignalR
+    capture: SignalRW
+
+
+class Capture(str, Enum):
+    No = "No"
+    Value = "Value"
+    Diff = "Diff"
+    Sum = "Sum"
+    Mean = "Mean"
+    Min = "Min"
+    Max = "Max"
+    MinMax = "Min Max"
+    MinMaxMean = "Min Max Mean"
+
+
+def get_capture_signals(
+    panda: Device, path_prefix: Optional[str] = ""
+) -> Dict[str, SignalR]:
+    """Get dict mapping a capture signal's name to the signal itself"""
+    # TODO makecommon with code from device_save_loader?
+    if not path_prefix:
+        path_prefix = ""
+    signals: Dict[str, SignalR[Any]] = {}
+    for attr_name, attr in panda.children():
+        dot_path = f"{path_prefix}{attr_name}"
+        if isinstance(attr, SignalR) and attr_name.endswith("_capture"):
+            signals[dot_path] = attr
+        attr_signals = get_capture_signals(attr, path_prefix=dot_path + ".")
+        signals.update(attr_signals)
+    return signals
+
+
+# This should return a dictionary which maps to a dict which contains the Capture
+# signal object,
+# and the value of that signal
+async def get_signals_marked_for_capture(
+    capture_signals: Dict[str, SignalR]
+) -> Dict[str, Union[SignalR, Capture]]:
+    # Read signals to see if they should be captured
+
+    do_read = [signal.get_value() for signal in capture_signals.values()]
+
+    signal_values = await asyncio.gather(*do_read)
+
+    assert len(signal_values) == len(
+        capture_signals.keys()
+    ), "Length of read signals are different to length of signals"
+    signals_to_capture: Dict[str, Dict[str, SignalR | Capture]] = {}
+    for signal_path, signal_object, signal_value in zip(
+        capture_signals.keys(), capture_signals.values(), signal_values
+    ):
+        if (signal_value.value in iter(Capture)) and (signal_value.value != Capture.No):
```

Comment:
I'm wondering why this works, as the `Capture` enum defined above is not the same as the generated Enum subclass that the "fill in signals" code uses. Maybe because they both subclass str?